### PR TITLE
[BEAM-812] gcp-io: add ApiSurfaceTest

### PR DESCRIPTION
**Pull request by dhalperi**
Copied from SDK, but I manually recreated/verified the whitelist.

Surely there are refactoring improvements to be had so that we can more easily
reuse ApiSurface across modules.

R: @kennknowles 
R: @davorbonaci

### CODE REVIEW

**Comment by kennknowles** [comment]
Seems here that you _are_ reusing `ApiSurface` across modules, no?

**Comment by kennknowles** [comment]
If you want to use the list of modules that are on the SDK's surface, seems something like this would do it:

```
public class SdkApiSurface {
  public List<String> publicPackagesOnApiSurface = ...
}
```

It used to be that we needed more complex rules than just packages to manage the surface, but now I think it is just that simple.

**Comment by dhalperi** [comment]
Some desiderata for reuse:

* test automatically infers the module on which it should seed the API search.
* only 2 things in the file -- a whitelist and a ~1-line test that passes the whitelist as a parameter to some helper.
* test fails if there are things in the whitelist that are not exposed.

**Comment by kennknowles** [comment]
Those all sound like good starter JIRAs.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/ApiSurfaceTest.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    List<String> disallowedMessages = Lists.newArrayList();
+    for (Map.Entry<Class<?>, List<Class<?>>> entry : disallowedClasses.entrySet()) {
+      disallowedMessages.add(entry.getKey() + " exposed via:\n\t\t"
+      + Joiner.on("\n\t\t").join(entry.getValue()));
+    }
+    Collections.sort(disallowedMessages);
+
+    if (!disallowedMessages.isEmpty()) {
+      fail("The following disallowed classes appear in the public API surface of the SDK:\n\t"
```
Comment (kennknowles) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/ApiSurfaceTest.java]: nit: might be clearer that the above is a no-op if the set of messages is only constructed inside here, where it is going to be used.

**Comment by davorbonaci** [comment]
LGTM. (The desired state would be really awesome; let's make sure we have JIRAs for those.)

Some things in the allow list are suspicious on the first look: sun, junit, etc.

**Comment by dhalperi** [comment]
Filed https://issues.apache.org/jira/browse/BEAM-878

Improved the exclusions of tests and test-related structures , which let me remove more things from the whitelist. PTAL.

**Pull request closed without merging**